typescript/infra/scripts/warp-routes/monitor-warp-routes-balances.ts

```diff
@@ -240,4 +252,146 @@ export function updateTokenBalanceMetrics(
   });
 }
 
+export function updateXERC20LimitsMetrics(
+  xERC20Limits: {
+    chain: string;
+    mint: number;
+    burn: number;
+    mintMax: number;
+    burnMax: number;
+  }[],
+) {
+  xERC20Limits.forEach((limit) => {
+    xERC20LimitsGauge
+      .labels({
+        chain_name: limit.chain,
+        limit_type: 'mint',
+      })
+      .set(limit.mint);
+    xERC20LimitsGauge
+      .labels({
+        chain_name: limit.chain,
+        limit_type: 'burn',
+      })
+      .set(limit.burn);
+    xERC20LimitsGauge
+      .labels({
+        chain_name: limit.chain,
+        limit_type: 'mintMax',
+      })
+      .set(limit.mintMax);
+    xERC20LimitsGauge
+      .labels({
+        chain_name: limit.chain,
+        limit_type: 'burnMax',
+      })
+      .set(limit.burnMax);
+    logger.info('xERC20 limits updated for chain', {
+      chain: limit.chain,
+      mint: limit.mint,
+      burn: limit.burn,
+      mintMax: limit.mintMax,
+      burnMax: limit.burnMax,
+    });
+  });
+}
+
+function getXerc20Limits(
+  warpCoreConfig: WarpCoreConfig,
+  multiProvider: MultiProvider,
+): Promise<
+  {
+    chain: string;
+    mint: number;
+    mintMax: number;
+    burn: number;
+    burnMax: number;
+  }[]
+> {
+  const result = Promise.all(
+    warpCoreConfig.tokens
+      .filter(
+        (t) =>
+          t.standard === TokenStandard.EvmHypXERC20 ||
+          t.standard === TokenStandard.EvmHypXERC20Lockbox,
+      )
+      .map(async (t) => {
+        const provider = multiProvider.getProvider(t.chainName);
+        const router = t.addressOrDenom!;
+        const xerc20Address =
+          t.standard === TokenStandard.EvmHypXERC20Lockbox
+            ? await HypXERC20Lockbox__factory.connect(router, provider).xERC20()
+            : await HypXERC20__factory.connect(router, provider).wrappedToken();
+
+        const xerc20 = IXERC20__factory.connect(xerc20Address, provider);
+        const mintCurrent = await xerc20.mintingCurrentLimitOf(router);
+        const mintMax = await xerc20.mintingMaxLimitOf(router);
+        const burnCurrent = await xerc20.burningCurrentLimitOf(router);
+        const burnMax = await xerc20.burningMaxLimitOf(router);
+
+        return {
+          chain: t.chainName,
+          mint: parseFloat(ethers.utils.formatUnits(mintCurrent)),
+          mintMax: parseFloat(ethers.utils.formatUnits(mintMax)),
+          burn: parseFloat(ethers.utils.formatUnits(burnCurrent)),
+          burnMax: parseFloat(ethers.utils.formatUnits(burnMax)),
+        };
+      }),
+  );
+
+  return result;
+}
+
+async function checkXERC20Limits(checkFrequency: number) {
+  const registry = await getRegistry();
+  const symbol = 'EZETH';
```

Comment:
A couple things about this version:
- doesn't work for things other than EZETH - imo we should make this configurable before merging
- breaks if there is more than one EZETH deployment

In the future, I wonder if we should make use of warp route IDs, which were added kinda recently to identify specific warp route deployments.

Some context on the IDs:
- basically option 1 as described here https://discord.com/channels/935678348330434570/1240001006050213918/1242481517352063038
- implemented here https://github.com/hyperlane-xyz/hyperlane-registry/pull/43/files#diff-a7ae5c67bef044df975c964ec7c0b32839ae904c5c669de9ae55dfb0f8674e87R62 and https://github.com/hyperlane-xyz/hyperlane-registry/pull/43/files#diff-b8ce851525e7bc41ba0aaec5c12987e029ae77a4fd9c0c60d81c766ee73130e1R54-R57

Iiuc the format of a warp ID is `${token}/${sortedChains.join('-')}`, e.g. `EZETH/arbitrum-base-blast-bsc-ethereum-linea-mode-optimism` (can be found by looking at the names / path of https://github.com/hyperlane-xyz/hyperlane-registry/tree/main/deployments/warp_routes/EZETH)

We could pass the warp ID as a flag to the deploy script that ensures the warp ID exists, and then that could pass the warp ID as a flag to this script via helm.

However this does give us a couple issues:
- these IDs are super verbose and cumbersome to write / remember
- atm the name of the infra file that's passed into the deploy script with the config is used to figure out what to name the helm / k8s deployment and resources (https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/typescript/infra/scripts/warp-routes/helm.ts#L25). moving to using the warp ID instead won't work bc the resource name requirements in k8s are pretty conservative - `EZETH/arbitrum-base-blast-bsc-ethereum-linea-mode-optimism` would def be too long and it'd complain about caps and `/`. So we'd need to figure out a different naming setup
- It's a little awkward having the xerc20 path involve the registry and the other non-xerc20 path not involve the registry
- Honestly am not certain if warp IDs are a notion that will stick around long or not. We can get an indication from the product team here tho

So I see our options as the following, in rough order of lowest lift -> highest lift:
1. Just leave this as-is, don't accommodate any other xERC20s other than EZETH
2. leave mostly as-is, but get 'EZETH' here from `EZETH-deployments.yaml` so it's a little more dynamic / configurable
3. kick the can further - for now just shamelessly duplicate and copy https://github.com/hyperlane-xyz/hyperlane-registry/blob/main/deployments/warp_routes/EZETH/arbitrum-base-blast-bsc-ethereum-linea-mode-optimism-config.yaml into `EZETH-deployments.yaml` here in infra (with the appropriate schema). We don't get anything from the registry, just the file passed in. Then later we can revisit to do something more "correct" which is probably close to what I describe above
4. Implement something close to what I describe above now, knowing that we'll have deviating paths for xerc20 vs not, and accepting we'll have to make some changes to our infra setup to get this working
5. full on commit to moving to the registry for everything with a setup similar to what I describe above for both the xerc20 and erc20 path

I'm kinda partial to (3) to get this merged and over the line with minimal effort while still being configurable / usable for other tokens in the future, and then we schedule follow up work to do (5) in a way that feels right. But am open to chatting about all these, I don't feel very strongly

Reply:
I guess if we do 5 eventually, 3 now feels like a move and work in the wrong direction